crates/sui-graphql-rpc/schema.graphql

```diff
@@ -0,0 +1,840 @@
+}
+
+type Subscription {
+  events(filter: EventFilter): Event
+  transactions(filter: TransactionBlockFilter): TransactionBlock
+}
+
+# String containing 32B hex-encoded address
+scalar SuiAddress
+
+# String containing 32B Base58-encoded BLAKE2b256 digest
+scalar Digest
+
+# String representation of a 64 bit unsigned integer
+scalar U64
+
+# String representation of a 128 bit signed integer
+scalar I128
+
+# String representation of a 128 bit unsigned integer
+scalar U128
+
+# String containing Base58-encoded binary data.
+scalar Bytes
+
+scalar JSON
+
+# Fully qualified concrete type instantiation, with a flat
+# representation:
+#
+# ([package]::[module]::)?[name](<[param], ...>)?.
+scalar TypeInst
+
+# Move type from a function (parameter or return type) or struct
+# (field) signature.  May contain type parameters, referred to by
+# index.
+scalar TypeSignature
+
+# Filter either by the digest, or the sequence number, or neither, to
+# get the latest checkpoint.
+input CheckpointFilter {
+  digest: Digest
+  sequenceNumber: U64
+}
+
+enum Order { ASC, DESC }
+
+input ObjectQuery {
+  # one of these
+  any: [ObjectQuery]
+  all: [ObjectQuery]
+  not: ObjectQuery
+
+  # (x)or any of these
+
+  # Cascading (type requires module requires package)
+  package: SuiAddress
+  module: String
+  type: String
+
+  owner: SuiAddress
+  objectId: SuiAddress
+  version: U64
+}
+
+input EventFilter {
+  # one of these
+  any: [EventFilter]
+  all: [EventFilter]
+  not: EventFilter
+
+  # (x)or any of these
+  sender: SuiAddress
+  transaction: Digest
+
+  # Cascading (module requires package)
+  emittingPackage: SuiAddress
+  emittingModule: String
+
+  # Cascading (type requires module requires package)
+  eventPackage: SuiAddress
+  eventModule: String
+  eventType: String
+
+  eventField: EventField
+
+  startTimeMs: U64
+  endTimeMs: U64
+}
+
+input EventField {
+  path: String!
+  value: JSON!
+}
+
+input TransactionBlockFilter {
+  # one of these
+  any: [TransactionBlockFilter]
+  all: [TransactionBlockFilter]
+  not: TransactionBlockFilter
+
+  # (x)or any of these
+
+  # Cascading (function requires module requires package)
+  package: SuiAddress
+  module: String
+  function: String
+
+  kind: TransactionBlockKindInput
+  checkpoint: U64
+  fromAddress: SuiAddress
+  toAddress: SuiAddress
+
+  inputObject: SuiAddress
+  changedObject: SuiAddress
+}
+
+type ConsistentRange {
+  first: Checkpoint
+  last: Checkpoint
+}
+
+interface Owner {
+  id: SuiAddress!
+
+  objects(query: ObjectQuery): ObjectQueryConnection
+
+  balance(type: TypeInst!): Balance
+  balances: BalanceConnection
+
+  coins(type: TypeInst): CoinConnection
+
+  stakes: StakeConnection
+
+  defaultNameServiceName: String
+  nameServiceNames: NameServiceNamesConnection
+}
+
+# Returned by Object.owner, where we can't disambiguate between
+# Address and Object.
+type AmbiguousOwner implements Owner {}
+
+type Address implements Owner {
+  transactionBlocks(
```

Comment:
I'm actually curious if this can live on the `Owner` as well?

Reply:
Only the received transaction blocks apply in both cases (which is why currently I ported only that one over).

We could make the interface general but have objects always return empty responses for SENT, and PAID.

FROM is a weird one because technically you can never send from an object -- you send to an object, the object's (transitive) owner address receives it and then can send it on, but if it all happens in a single transaction, it's going to look like it was sent from the object.